tools/config/__init__.py

```diff
@@ -1058,9 +1062,31 @@ def validate_config(self):
 
         Arguments: None
         """
+
         params, _ = self.get_config_data()
+        err_msg = ""
+
+        for name, param in sorted(params.items()):
+            min      = param.value_min
+            max      = param.value_max
+            accepted = param.accepted_values
+            value    = param.value
+
+            if (value < min or (value > max if max is not None else False)):
+                err_msg += "\nInvalid config range for %s, is not in the required range: [%s:%s]"\
+                                % (param, 
+                                   min if min is not None else "-inf", 
+                                   max if max is not None else "inf")
+
+            if accepted and value not in accepted:
+                err_msg += "\nInvalid config range for %s, is not an accepted value: %s"\
+                                % (param, accepted)
```

Comment:
Would it make sense to make the `accepted_range` and the `value_min`/`value_max` cases exclusive? I think we'd want to raise a `ConfigException` if both a range and a min/max were specified right?

----

Also interesting note, this error message is for saying "Choices x, y, and z are valid, select from that list please" right? When its printed out, it appears as:

```
Invalid config range for option, is not an accepted value: [0, 512]
```

The syntax `[0, 512]` is actually used in math to denote an **interval** (here we call it a range) of numbers: https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints. In this case, the range would be 0 - 512 (inclusive of 0 and 512), vs the intended "pick 0 **or** 512". Based on your other notes, it looks like we're denoting ranges as [min:max]. So maybe not a big deal, but I admit I got a bit confused!

Reply:
If you got confused that's a bad sign for sure, I can reword that to make it more clear. Making range and accepted values exclusive is a really good point.